### PR TITLE
Gen3: XFeat experiment.

### PR DESCRIPTION
**Pull request by kkeroo**
This PR adds an experiment with `XFeat` model from HubAI. We have two modes available: `mono` and `stereo`. Because the visualization is pretty specific - we draw matches (lines) between two frames - we can not visualize it nicely in our visualizer so we draw visualization with cv2 using custom node and send the complete `dai.ImgFrame` to the Visualizer.

In requirements we install dai-nodes with git because I discovered some bug in the XFeat parser.

The experiment is in new folder `image-matching`. Does it make sense to rather have it in the `neural-stereo-matching`?

The README is missing the demo GIF. Where do we store it?

### CODE REVIEW

**Review by klemen1999 — APPROVED** [review]

**Comment by klemen1999** [comment]
> The experiment is in new folder image-matching. Does it make sense to rather have it in the neural-stereo-matching?

I think `image-matching` makes sense for now, we can change later if needed

> The README is missing the demo GIF. Where do we store it?

Let's create a "media" folder and put it there. And then let's create a new entry in the main README [here](https://github.com/luxonis/depthai-experiments/tree/xfeat-gen3/gen3/neural-networks). It would be under title "Advanced examples" and subtitle "Image matching". And there we would have same style of table as we do for generic examples and list HubAI models used in the experiment, gen3 link, gen2 link (if it exists), if it works for RVC2 and RVC4 and finally a link to the gif (let's generate one on the stereo example)

**Comment by kkeroo** [comment]
> Let's create a "media" folder and put it there. And then let's create a new entry in the main README [here](https://github.com/luxonis/depthai-experiments/tree/xfeat-gen3/gen3/neural-networks). It would be under title "Advanced examples" and subtitle "Image matching". And there we would have same style of table as we do for generic examples and list HubAI models used in the experiment, gen3 link, gen2 link (if it exists), if it works for RVC2 and RVC4 and finally a link to the gif (let's generate one on the stereo example)

Hm, I am thinking if repo size wouldn't grow too much by storing all GIFs in the repo. What if we store it on G. Drive or somewhere?

**Comment by klemen1999** [comment]
> Hm, I am thinking if repo size wouldn't grow too much by storing all GIFs in the repo. What if we store it on G. Drive or somewhere?

I don't think that storing them in the repo should be that big of an issue, we likely won't have it for each model. But if it becomes an issue then we can maybe open an Issue on the repo intended to just pasting gifs/images and then we reference this in READMEs

**Comment by jkbmrz** [comment]
I also like the idea of putting it under image-matching folder!

> I am thinking if repo size wouldn't grow too much by storing all GIFs in the repo.

Repo size is already quite big so this should have much effect. However, we can reduce the size of the GIFs by skipping frames /decreasing their shape

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+## Usage
+
+```
+python3 main.py --model <MODEL> --device <DEVICE> --fps_limit <FPS_LIMIT>
+```
+
+
+- `<MODEL>`: Model slug from Luxonis HubAI. Default: `luxonis/xfeat:mono-240x320`.
```
Comment (jkbmrz) [inline gen3/neural-networks/advanced-examples/image-matching/xfeat/README.md:31]: Rename `Model slug` -> `HubAI Model Reference` (this is how we changed in it generic example)

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/mono.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        visualizer.addTopic("Matches", custom_visualizer.out, "images")
+
+        print("Pipeline created.")
+        print("\nPress 's' to set the reference frame.\n")
+
+        pipeline.start()
+        visualizer.registerPipeline(pipeline)
+
+        while pipeline.isRunning():
+            time.sleep(1 / 30)
```
Comment (jkbmrz) [inline gen3/neural-networks/advanced-examples/image-matching/xfeat/mono.py:39]: is this still needed even if monitoring the key press?

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/utils/arguments.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        help="FPS limit for the model runtime.",
+        required=False,
+        default=30,
+        type=int,
+    )
+
+    parser.add_argument(
+        "-m",
+        "--model",
+        help="The XFeat model. Get it from the Luxonis HubAI.",
```
Comment (jkbmrz) [inline gen3/neural-networks/advanced-examples/image-matching/xfeat/utils/arguments.py:28]: maybe mention one needs to provide the HubAI model reference (ex-slug)

**Comment by kkeroo** [comment]
Okay, I added the GIF (~2MB) to the media folder.

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/utils/arguments.py (hunk 3 above)
Comment (kkeroo) [reply gen3/neural-networks/advanced-examples/image-matching/xfeat/utils/arguments.py]: Fixed in [2cb2f3e](https://github.com/luxonis/depthai-experiments/pull/562/commits/2cb2f3e6e90caf1f605cb618cdbe14ac89197f66).

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/mono.py (hunk 2 above)
Comment (kkeroo) [reply gen3/neural-networks/advanced-examples/image-matching/xfeat/mono.py]: Fixed in [61db702](https://github.com/luxonis/depthai-experiments/pull/562/commits/61db7023039ca7d02e4e8db234b44b6aea7d0f72).

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/main.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import depthai as dai
+from utils.arguments import initialize_argparser
+from stereo import stereo_mode
+from mono import mono_mode
+
+_, args = initialize_argparser()
+
+
+model_slug = args.model
```
[the remark this answers is not in the document]
Comment (kkeroo) [reply gen3/neural-networks/advanced-examples/image-matching/xfeat/main.py]: Fixed in [6a13bcf](https://github.com/luxonis/depthai-experiments/pull/562/commits/6a13bcf12caa9601e10c84b69b8ecb2e7a50b78e).

File: gen3/neural-networks/advanced-examples/image-matching/xfeat/README.md (hunk 1 above)
Comment (kkeroo) [reply gen3/neural-networks/advanced-examples/image-matching/xfeat/README.md]: Fixed in [6a13bcf](https://github.com/luxonis/depthai-experiments/pull/562/commits/6a13bcf12caa9601e10c84b69b8ecb2e7a50b78e).

**Review by klemen1999 — APPROVED** [review]

**Pull request merged**